File: testsuite/python/particle_slice.py
# Copyright (C) 2010-2019 The ESPResSo project
#
# This file is part of ESPResSo.
#
# ESPResSo is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# ESPResSo is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import unittest as ut
import unittest_decorators as utx
import espressomd
import numpy as np


class ParticleSliceTest(ut.TestCase):

    system = espressomd.System(box_l=3 * [10])

    def setUp(self):
        self.fix_flags = [[False, False, False], [False, False, True]]
        self.p0 = self.system.part.add(pos=[0, 0, 0])
        self.p1 = self.system.part.add(pos=[0, 0, 1])
        self.p2 = self.system.part.add(pos=[0, 0, 2])
        self.p3 = self.system.part.add(pos=[0, 0, 3])

        self.all_partcls = self.system.part.all()
        self.p0p1 = self.system.part.by_ids([0, 1])
        self.p2p3 = self.system.part.by_ids([2, 3])

        if espressomd.has_features(["EXTERNAL_FORCES"]):
            self.p1.fix = self.fix_flags[1]
            np.testing.assert_array_equal(
                np.copy(self.p0.fix), self.fix_flags[0])
            np.testing.assert_array_equal(
                np.copy(self.p1.fix), self.fix_flags[1])
            np.testing.assert_array_equal(
                np.copy(self.p0p1.fix), self.fix_flags)

    def tearDown(self):
        self.system.part.clear()

    @utx.skipIfMissingFeatures(["EXTERNAL_FORCES"])
    def test_1_set_different_values(self):
        self.fix_flags[0] = [1, 0, 0]
        self.fix_flags[1] = [1, 0, 0]
        self.p0p1.fix = self.fix_flags
        np.testing.assert_array_equal(
            np.copy(self.p0p1.fix), self.fix_flags)

    @utx.skipIfMissingFeatures(["EXTERNAL_FORCES"])
    def test_2_set_same_value(self):
        self.fix_flags[0] = [0, 1, 0]
        self.fix_flags[1] = [0, 1, 0]
        self.p0p1.fix = self.fix_flags[1]
        np.testing.assert_array_equal(
            np.copy(self.p0p1.fix), self.fix_flags)

    @utx.skipIfMissingFeatures(["EXTERNAL_FORCES"])
    def test_3_set_one_value(self):
        self.fix_flags[1] = [0, 0, 1]
        self.p1.fix = self.fix_flags[1]
        np.testing.assert_array_equal(
            np.copy(self.p0p1.fix), self.fix_flags)

    @utx.skipIfMissingFeatures(["EXTERNAL_FORCES"])
    def test_4_str(self):
        self.assertEqual(repr(self.p0.fix),
                         repr(np.array([0, 0, 0])))
        self.assertEqual(repr(self.p0p1.fix),
                         repr(np.array([[0, 0, 0], [0, 0, 1]])))

    def test_pos_str(self):
        self.p0.pos = [0, 0, 0]
        self.p1.pos = [0, 0, 1]
        self.assertEqual(repr(self.p0.pos),
                         repr(np.array([0.0, 0.0, 0.0])))
        self.assertEqual(repr(self.p0p1.pos),
                         repr(np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])))

    @utx.skipIfMissingFeatures(["ELECTROSTATICS"])
    def test_scalar(self):
        self.p0.q = 1.3
        self.assertEqual(self.p0.q, 1.3)
        self.p0p1.q = 2.0
        self.assertEqual(self.p0.q, 2)
        self.assertEqual(self.p1.q, 2)
        self.p0p1.q = 3
        self.assertEqual(self.p0.q, 3)
        self.assertEqual(self.p1.q, 3)
        self.p0p1.q = [-1, 1.0]
        self.assertEqual(self.p0.q, -1)
        self.assertEqual(self.p1.q, 1)
        qs = self.p0p1.q
        self.assertEqual(qs[0], -1)
        self.assertEqual(qs[1], 1)

    def test_bonds(self):

        fene = espressomd.interactions.FeneBond(k=1, d_r_max=1, r_0=1)
        self.system.bonded_inter.add(fene)

        # Setter

        # tuple
        b = fene, 0
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])

        # list
        self.all_partcls.bonds = []
        b = [fene, 0]
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])

        # nested list single
        self.all_partcls.bonds = []
        b = [[fene, 0]]
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])

        # nested list multi
        self.all_partcls.bonds = []
        b = [[fene, 0], [fene, 1]]
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ()])

        # nested tuple single
        self.all_partcls.bonds = []
        b = ((fene, 0),)
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])

        # nested tuple multi
        self.all_partcls.bonds = []
        b = ((fene, 0), (fene, 1))
        self.p2.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ()])

        # Add/Del bonds
        self.all_partcls.bonds = []
        self.p2.add_bond((fene, 0))
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])
        self.p2.add_bond((fene, 1))
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ()])
        self.p2.delete_bond((fene, 1))
        self.assertEqual(self.all_partcls.bonds, [(), (), ((fene, 0),), ()])
        self.p2.delete_bond((fene, 0))
        self.assertEqual(self.all_partcls.bonds, [(), (), (), ()])

        self.all_partcls.bonds = []
        self.p2.add_bond((fene, 0))
        self.p2.add_bond((fene, 1))
        self.p2.delete_all_bonds()
        self.assertEqual(self.all_partcls.bonds, [(), (), (), ()])

        # Slices

        # tuple for all
        self.all_partcls.bonds = []
        b = fene, 0
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 0),)])

        # list for all
        self.all_partcls.bonds = []
        b = [fene, 0]
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 0),)])

        # nested list single for all
        self.all_partcls.bonds = []
        b = [[fene, 0]]
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 0),)])

        # nested list multi for all
        self.all_partcls.bonds = []
        b = [[fene, 0], [fene, 1]]
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ((fene, 0), (fene, 1))])

        # tuples for each
        self.all_partcls.bonds = []
        b = (((fene, 0),), ((fene, 1),))
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 1),)])

        # lists for each
        self.all_partcls.bonds = []
        b = [[[fene, 0]], [[fene, 1]]]
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 1),)])

        # multi tuples for each
        self.all_partcls.bonds = []
        b = (((fene, 0), (fene, 1)), ((fene, 0), (fene, 1)))
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ((fene, 0), (fene, 1))])

        # multi lists for each
        self.all_partcls.bonds = []
        b = [[[fene, 0], [fene, 1]], [[fene, 0], [fene, 1]]]
        self.p2p3.bonds = b
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ((fene, 0), (fene, 1))])

        # Add/Del bonds
        self.all_partcls.bonds = []
        self.p2p3.add_bond((fene, 0))
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 0),)])
        self.p2p3.add_bond((fene, 1))
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0), (fene, 1)), ((fene, 0), (fene, 1))])
        self.p2p3.delete_bond((fene, 1))
        self.assertEqual(self.all_partcls.bonds,
                         [(), (), ((fene, 0),), ((fene, 0),)])
        self.p2p3.delete_bond((fene, 0))
        self.assertEqual(self.all_partcls.bonds, [(), (), (), ()])

        b = [[[fene, 0], [fene, 1]], [[fene, 0], [fene, 1]]]
        self.p2p3.bonds = b
        self.all_partcls.delete_all_bonds()
        self.assertEqual(self.all_partcls.bonds, [(), (), (), ()])

    @utx.skipIfMissingFeatures(["EXCLUSIONS"])
    def test_exclusions(self):

        def assert_exclusions_equal(B):
            A = self.all_partcls.exclusions
            self.assertEqual([x.tolist() for x in A], B)

        # Setter
        # int
        self.all_partcls.exclusions = []
        b = 1
        self.p2.exclusions = b
        assert_exclusions_equal([[], [2], [1], []])

        # single list
        self.all_partcls.exclusions = []
        b = [1]
        self.p2.exclusions = b
        assert_exclusions_equal([[], [2], [1], []])

        # tuple
        self.all_partcls.exclusions = []
        b = (0, 1)
        self.p2.exclusions = b
        assert_exclusions_equal([[2], [2], [0, 1], []])

        # list
        self.all_partcls.exclusions = []
        b = [0, 1]
        self.p2.exclusions = b
        assert_exclusions_equal([[2], [2], [0, 1], []])

        # Add/Del exclusions
        self.all_partcls.exclusions = []
        self.p2.add_exclusion(1)
        assert_exclusions_equal([[], [2], [1], []])
        self.p2.add_exclusion(0)
        assert_exclusions_equal([[2], [2], [1, 0], []])
        self.p2.delete_exclusion(0)
        assert_exclusions_equal([[], [2], [1], []])
        self.p2.delete_exclusion(1)
        assert_exclusions_equal([[], [], [], []])

        # Slices

        # single list for all
        self.all_partcls.exclusions = []
        b = [1]
        self.p2p3.exclusions = b
        assert_exclusions_equal([[], [2, 3], [1], [1]])

        # list for all
        self.all_partcls.exclusions = []
        b = [0, 1]
        self.p2p3.exclusions = b
        assert_exclusions_equal([[2, 3], [2, 3], [0, 1], [0, 1]])

        # single list for each
        self.all_partcls.exclusions = []
        b = [[0], [0]]
        self.p2p3.exclusions = b
        assert_exclusions_equal([[2, 3], [], [0], [0]])

        # multi list for each
        self.all_partcls.exclusions = []
        b = [[0, 1], [0, 1]]
        self.p2p3.exclusions = b
        assert_exclusions_equal([[2, 3], [2, 3], [0, 1], [0, 1]])

        # Add/Del exclusions
        self.all_partcls.exclusions = []
        self.p2p3.add_exclusion(1)
        assert_exclusions_equal([[], [2, 3], [1], [1]])
        self.p2p3.add_exclusion(0)
        assert_exclusions_equal([[2, 3], [2, 3], [1, 0], [1, 0]])
        self.p2p3.delete_exclusion(0)
        assert_exclusions_equal([[], [2, 3], [1], [1]])
        self.p2p3.delete_exclusion(1)
        assert_exclusions_equal([[], [], [], []])

    @utx.skipIfMissingFeatures("VIRTUAL_SITES_RELATIVE")
    def test_vs_relative(self):

        self.system.part.clear()
        p0 = self.system.part.add(pos=[0, 0, 0])
        self.system.part.add(pos=3 * [[0, 0, 0]])
        p0.vs_relative = [1, 1.0, (1.0, 1.0, 1.0, 1.0)]
        all_partcls = self.system.part.all()

        self.assertEqual(repr(all_partcls.vs_relative),
                         repr([(1, 1.0, np.array([1., 1., 1., 1.])),
                               (0, 0.0, np.array([1., 0., 0., 0.])),
                               (0, 0.0, np.array([1., 0., 0., 0.])),
                               (0, 0.0, np.array([1., 0., 0., 0.]))]))

        all_partcls.vs_relative = [1, 1.0, (1.0, 1.0, 1.0, 1.0)]

        self.assertEqual(repr(all_partcls.vs_relative),
                         repr([(1, 1.0, np.array([1., 1., 1., 1.])),
                               (1, 1.0, np.array([1., 1., 1., 1.])),
                               (1, 1.0, np.array([1., 1., 1., 1.])),
                               (1, 1.0, np.array([1., 1., 1., 1.]))]))

        all_partcls.vs_relative = [[1, 1.0, (1.0, 1.0, 1.0, 1.0)],
                                   [1, 2.0, (1.0, 1.0, 1.0, 1.0)],
                                   [1, 3.0, (1.0, 1.0, 1.0, 1.0)],
                                   [1, 4.0, (1.0, 1.0, 1.0, 1.0)]]

        self.assertEqual(repr(all_partcls.vs_relative),
                         repr([(1, 1.0, np.array([1., 1., 1., 1.])),
                               (1, 2.0, np.array([1., 1., 1., 1.])),
                               (1, 3.0, np.array([1., 1., 1., 1.])),
                               (1, 4.0, np.array([1., 1., 1., 1.]))]))

    def test_multiadd(self):
        self.system.part.clear()
        positions = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
        partcls_multiadd = self.system.part.add(pos=positions)
        for p in positions:
            self.system.part.add(pos=p)
        partcls_singleadd = self.system.part.by_ids(range(3, 6))
        np.testing.assert_allclose(
            np.copy(partcls_multiadd.pos), np.copy(partcls_singleadd.pos))

        with self.assertRaises(ValueError):
            self.system.part.add(pos=([1, 1, 1], [2, 2, 2]), type=0)
        with self.assertRaises(ValueError):
            self.system.part.add(pos=([1, 1, 1], [2, 2, 2]), type=(0, 1, 2))

        self.system.part.clear()
        p0, p1 = self.system.part.add(pos=([1, 1, 1], [2, 2, 2]), type=(0, 1))
        self.assertEqual(p0.type, 0)
        self.assertEqual(p1.type, 1)

    def test_empty(self):
        np.testing.assert_array_equal(
            self.system.part.by_ids(
                []).pos, np.empty(0))

    def test_len(self):
        self.assertEqual(len(self.system.part.by_ids([])), 0)
        self.assertEqual(len(self.system.part.by_ids([0])), 1)
        self.assertEqual(len(self.system.part.by_ids([0, 1, 2])), 3)

    def test_non_existing_property(self):
        with self.assertRaises(AttributeError):
            self.all_partcls.thispropertydoesnotexist = 1.0


if __name__ == "__main__":
    ut.main()

File: src/script_interface/particle_data/ParticleSlice.cpp
/*
 * Copyright (C) 2022-2026 The ESPResSo project
 *
 * This file is part of ESPResSo.
 *
 * ESPResSo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ESPResSo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "ParticleSlice.hpp"
#include "ParticleHandle.hpp"

#include "core/bonds.hpp"
#include "core/nonbonded_interactions/nonbonded_interaction_data.hpp"
#include "core/particle_node.hpp"
#include "core/system/System.hpp"

#include "script_interface/Context.hpp"
#include "script_interface/Exception.hpp"
#include "script_interface/Variant.hpp"
#include "script_interface/get_value.hpp"

#include <utils/Vector.hpp>
#include <utils/mpi/gather_buffer.hpp>

#include <algorithm>
#include <functional>
#include <iterator>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <variant>
#include <vector>

namespace ScriptInterface {
namespace Particles {

static void set_particles_bonds(
    std::vector<int> const &pids,
    std::vector<std::vector<int>> const &all_bonds_ids,
    std::vector<std::vector<std::vector<int>>> const &all_bonds_partner_ids,
    ::CellStructure &cell_structure, ::System::System &system) {
  for (std::size_t i = 0; i < pids.size(); ++i) {
    auto const pid = pids[i];
    auto const bonds_ids = all_bonds_ids[i];
    auto const bonds_partner_ids = all_bonds_partner_ids[i];
    // Remove old bonds
    auto p = cell_structure.get_local_particle(pid);
    if (p != nullptr and not p->is_ghost()) {
      p->bonds().clear();
    }
    // Add new bonds
    for (std::size_t j = 0; j < bonds_ids.size(); ++j) {
      std::vector<int> particle_ids = {pid};
      std::ranges::copy(bonds_partner_ids[j], std::back_inserter(particle_ids));
      ::add_bond(system, bonds_ids[j], particle_ids);
      system.on_particle_change();
    }
  }
}

#ifdef ESPRESSO_EXCLUSIONS
static void
set_particles_exclusions(std::vector<int> const &pids,
                         std::vector<std::vector<int>> const &exclusion_lists,
                         boost::mpi::communicator const &comm,
                         ::CellStructure &cell_structure,
                         ::System::System &system) {
  for (std::size_t i = 0; i < pids.size(); ++i) {
    auto const pid = pids[i];
    auto const &exclusion_list = exclusion_lists[i];
    for (auto const excluded_pid : exclusion_list) { // collective communication
      particle_exclusion_sanity_checks(pid, excluded_pid, cell_structure, comm);
    }
    auto p = cell_structure.get_local_particle(pid);
    if (p != nullptr and not p->is_ghost()) {
      // Remove all excluded ids of this particle
      for (auto const old_excluded_pid : p->exclusions()) {
        local_remove_exclusion(pid, old_excluded_pid, cell_structure);
      }
      // Add new excluded ids for this particle
      for (auto const excluded_pid : exclusion_list) {
        if (not p->has_exclusion(excluded_pid)) {
          local_add_exclusion(pid, excluded_pid, cell_structure);
        }
      }
    }
  }
  system.on_particle_change();
}
#endif // ESPRESSO_EXCLUSIONS

static void
set_particles_positions(std::vector<int> const &pids,
                        std::vector<Utils::Vector3d> const &positions) {
  for (std::size_t i = 0; i < pids.size(); ++i) {
    auto const pid = pids[i];
    auto const pos = positions[i];
    particle_checks(pid, pos);
    set_particle_pos(pid, pos);
  }
}

static void set_particles_types(std::vector<int> const &pids,
                                std::vector<int> const &types,
                                CellStructure &cell_structure,
                                ::System::System &system) {
  for (std::size_t i = 0; i < pids.size(); ++i) {
    auto const pid = pids[i];
    auto p = cell_structure.get_local_particle(pid);
    if (p != nullptr and not p->is_ghost()) {
      auto const old_type = p->type();
      auto const &new_type = types[i];
      if (new_type < 0) {
        throw std::domain_error(error_msg("type", "must be an integer >= 0"));
      }
      system.nonbonded_ias->make_particle_type_exist(new_type);
      on_particle_type_change(pid, old_type, new_type);
      p->type() = new_type;
    }
  }
}

#ifdef ESPRESSO_ELECTROSTATICS
static void set_particles_charges(std::vector<int> const &pids,
                                  std::vector<double> const &charges,
                                  CellStructure &cell_structure,
                                  ::System::System &system) {
  for (std::size_t i = 0; i < pids.size(); ++i) {
    auto const pid = pids[i];
    auto p = cell_structure.get_local_particle(pid);
    if (p != nullptr and not p->is_ghost()) {
      p->q() = charges[i];
    }
  }
  system.on_particle_charge_change();
}
#endif // ESPRESSO_ELECTROSTATICS

void ParticleSlice::do_construct(VariantMap const &params) {
  if (params.contains("__cell_structure")) {
    auto so = get_value<std::shared_ptr<CellSystem::CellSystem>>(
        params, "__cell_structure");
    so->configure(*this);
    m_cell_structure = so;
  }
  if (params.contains("__bonded_ias")) {
    m_bonded_ias = get_value<std::shared_ptr<Interactions::BondedInteractions>>(
        params, "__bonded_ias");
  }
  m_id_selection = get_value<std::vector<int>>(params, "id_selection");
  m_chunk_size = get_value_or<int>(params, "prefetch_chunk_size", 10000);
  if (not context()->is_head_node()) {
    return;
  }
  for (auto const pid : m_id_selection) {
    if (not particle_exists(pid)) {
      throw std::out_of_range("Particle does not exist: " +
                              std::to_string(pid));
    }
  }
}

Variant ParticleSlice::do_call_method(std::string const &name,
                                      VariantMap const &params) {
  if (name == "set_param_parallel") {
    auto const param_name = get_value<std::string>(params, "name");
    if (not params.contains("values")) {
      context()->parallel_try_catch([&]() {
        if (param_name == "bonds") {
          if (not params.contains("all_bonds_ids")) {
            throw Exception("Parameter 'all_bonds_ids' is missing");
          }
          if (not params.contains("all_bonds_partner_ids")) {
            throw Exception("Parameter 'all_bonds_partner_ids' is missing");
          }
        } else {
          throw Exception("Parameter 'values' is missing");
        }
      });
    }
    // Handle parameters with special setters
    if (m_special_parameters.contains(param_name)) {
      context()->parallel_try_catch([&]() {
        if (param_name == "pos") {
          set_particles_positions(
              m_id_selection,
              get_value<std::vector<Utils::Vector3d>>(params, "values"));
        } else if (param_name == "type") {
          set_particles_types(m_id_selection,
                              get_value<std::vector<int>>(params, "values"),
                              *get_cell_structure(), *get_system());
        }
#ifdef ESPRESSO_ELECTROSTATICS
        else if (param_name == "q") {
          std::vector<double> charges;
          if (is_type<std::vector<int>>(params.at("values"))) {
            auto tmp = get_value<std::vector<int>>(params, "values");
            charges = std::vector<double>(tmp.begin(), tmp.end());
          } else {
            charges = get_value<std::vector<double>>(params, "values");
          }
          set_particles_charges(m_id_selection, charges, *get_cell_structure(),
                                *get_system());

        }
#endif // ESPRESSO_ELECTROSTATICS
#ifdef ESPRESSO_EXCLUSIONS
        else if (param_name == "exclusions") {
          auto const excluded_pids =
              get_value<std::vector<std::vector<int>>>(params, "values");
          set_particles_exclusions(
              m_id_selection,
              get_value<std::vector<std::vector<int>>>(params, "values"),
              context()->get_comm(), *get_cell_structure(), *get_system());
        }
#endif // ESPRESSO_EXCLUSIONS
        else if (param_name == "bonds") {
          set_particles_bonds(
              m_id_selection,
              get_value<std::vector<std::vector<int>>>(params, "all_bonds_ids"),
              get_value<std::vector<std::vector<std::vector<int>>>>(
                  params, "all_bonds_partner_ids"),
              *get_cell_structure(), *get_system());
        }
      });
    } else {
      // Handle generic parameters
      context()->parallel_try_catch([&]() {
        std::visit(
            [&](auto &&vals) {
              set_from_vector_like(m_id_selection, param_name, vals, context(),
                                   m_cell_structure.lock(),
                                   m_bonded_ias.lock());
            },
            params.at("values"));
      });
    }
    return {};
  }
  if (name == "get_param_parallel") {
    auto const param_name = get_value<std::string>(params, "name");

    // handle special optimized properties
    if (param_name == "type") {
      auto const getter{[](Particle const &p) { return p.type(); }};
      return get_particles_properties<int>(m_id_selection, getter, context(),
                                           *get_cell_structure());
    }
    if (param_name == "q") {
      auto const getter{[](Particle const &p) { return p.q(); }};
      return get_particles_properties<double>(m_id_selection, getter, context(),
                                              *get_cell_structure());
    }
    if (param_name == "pos") {
      auto const &box_geo = *get_system()->box_geo;
      auto const getter = [&box_geo](Particle const &p) {
        return box_geo.unfolded_position(p.pos(), p.image_box());
      };
      return get_particles_properties<Utils::Vector3d>(
          m_id_selection, getter, context(), *get_cell_structure());
    }
    if (param_name == "pos_folded") {
      auto const &box_geo = *get_system()->box_geo;
      auto const getter = [&box_geo](Particle const &p) {
        return box_geo.folded_position(p.pos());
      };
      return get_particles_properties<Utils::Vector3d>(
          m_id_selection, getter, context(), *get_cell_structure());
    }

    // handle all other particle properties using expensive MPI reductions
    if (!context()->is_head_node()) {
      return {};
    }
    std::vector<Variant> result;
    result.reserve(m_id_selection.size());
    auto so = std::dynamic_pointer_cast<ParticleModifier>(
        context()->make_shared("Particles::ParticleModifier",
                               {{"id", -1},
                                {"__cell_structure", m_cell_structure.lock()},
                                {"__bonded_ias", m_bonded_ias.lock()}}));
    for (int pid : m_id_selection) {
      so->set_pid(pid);
      result.emplace_back(so->get_parameter(param_name));
    }
    return result;
  }

  if (not context()->is_head_node()) {
    return {};
  }
  if (name == "prefetch_particle_data") {
    auto p_ids = get_value<std::vector<int>>(params, "chunk");
    prefetch_particle_data(p_ids);
    return {};
  }
  if (name == "get_particle") {
    return context()->make_shared(
        "Particles::ParticleHandle",
        {{"id", get_value<int>(params, "p_id")},
         {"__cell_structure", m_cell_structure.lock()},
         {"__bonded_ias", m_bonded_ias.lock()}});
  }
  return {};
}

} // namespace Particles
} // namespace ScriptInterface
